staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -206,3 +218,125 @@ func (testVersioner) ObjectResourceVersion(obj runtime.Object) (uint64, error) {
+}
+
+type dummyStorage struct {
+	err error
+}
+
+type dummyWatch struct {
+	ch chan watch.Event
+}
+
+func (w *dummyWatch) ResultChan() <-chan watch.Event {
+	return w.ch
+}
+
+func (w *dummyWatch) Stop() {
+	close(w.ch)
+}
+
+func newDummyWatch() watch.Interface {
+	return &dummyWatch{
+		ch: make(chan watch.Event),
+	}
+}
+
+func (d *dummyStorage) Versioner() storage.Versioner { return nil }
+func (d *dummyStorage) Create(_ context.Context, _ string, _, _ runtime.Object, _ uint64) error {
+	return fmt.Errorf("unimplemented")
+}
+func (d *dummyStorage) Delete(_ context.Context, _ string, _ runtime.Object, _ *storage.Preconditions) error {
+	return fmt.Errorf("unimplemented")
+}
+func (d *dummyStorage) Watch(_ context.Context, _ string, _ string, _ storage.SelectionPredicate) (watch.Interface, error) {
+	return newDummyWatch(), nil
+}
+func (d *dummyStorage) WatchList(_ context.Context, _ string, _ string, _ storage.SelectionPredicate) (watch.Interface, error) {
+	return newDummyWatch(), nil
+}
+func (d *dummyStorage) Get(_ context.Context, _ string, _ string, _ runtime.Object, _ bool) error {
+	return fmt.Errorf("unimplemented")
+}
+func (d *dummyStorage) GetToList(_ context.Context, _ string, _ string, _ storage.SelectionPredicate, _ runtime.Object) error {
+	return d.err
+}
+func (d *dummyStorage) List(_ context.Context, _ string, _ string, _ storage.SelectionPredicate, listObj runtime.Object) error {
+	podList := listObj.(*example.PodList)
+	podList.ListMeta = metav1.ListMeta{ResourceVersion: "100"}
+	return d.err
+}
+func (d *dummyStorage) GuaranteedUpdate(_ context.Context, _ string, _ runtime.Object, _ bool, _ *storage.Preconditions, _ storage.UpdateFunc, _ ...runtime.Object) error {
+	return fmt.Errorf("unimplemented")
+}
+func (d *dummyStorage) Count(_ string) (int64, error) {
+	return 0, fmt.Errorf("unimplemented")
+}
+
+func TestListWithLimitAndRV0(t *testing.T) {
+	backingStorage := &dummyStorage{}
+	cacher, _ := newTestCacher(backingStorage, 0)
+	defer cacher.Stop()
+
+	pred := storage.SelectionPredicate{
+		Limit: 500,
+	}
+	result := &example.PodList{}
+
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+
+	// Inject error to underlying layer and check if cacher is not bypassed.
+	backingStorage.err = fmt.Errorf("dummy error")
+	err := cacher.List(context.TODO(), "pods/ns", "0", pred, result)
+	if err != nil {
+		t.Errorf("List with Limit and RV=0 should be served from cache: %v", err)
+	}
+}
+
+func TestGetToListWithLimitAndRV0(t *testing.T) {
+	backingStorage := &dummyStorage{}
+	cacher, _ := newTestCacher(backingStorage, 0)
+	defer cacher.Stop()
+
+	pred := storage.SelectionPredicate{
+		Limit: 500,
+	}
+	result := &example.PodList{}
+
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+
+	// Inject error to underlying layer and check if cacher is not bypassed.
+	backingStorage.err = fmt.Errorf("dummy error")
+	err := cacher.GetToList(context.TODO(), "pods/ns", "0", pred, result)
```

Comment:
same here, make a second call that hits storage to make sure our error is surfaced when storage is called